test/test_qat.py

```diff
@@ -123,17 +122,21 @@ def ref_op(x):
             else:
                 ref_op = compose([conv_op, bn_op, relu_op])
 
-            result_ref = ref_op(input)
-            result_actual = qat_op(input)
-            self.assertEqual(result_ref, result_actual)
+            input_clone = input.clone().detach_().requires_grad_()
+            for i in range(2):
+                result_ref = ref_op(input)
+                result_actual = qat_op(input_clone)
+                self.assertEqual(result_ref, result_actual)
 
-            # backward
-            dout = torch.randn(result_ref.size(), dtype=torch.float)
-            result_actual.backward(dout, retain_graph=True)
-            grad_ref = input.grad.cpu()
-            result_actual.backward(dout)
-            grad_actual = input.grad.cpu()
-            self.assertEqual(grad_ref, grad_actual)
+                # backward
+                dout = torch.randn(result_ref.size(), dtype=torch.float)
+                loss = (result_ref - dout).sum()
+                loss.backward()
+                grad_ref = input.grad.cpu()
+                loss = (result_actual - dout).sum()
+                loss.backward()
+                grad_actual = input_clone.grad.cpu()
```

Comment:
NIT: You dont need to create input_clone, you could set gradient to zero after you calculate grad_ref.
```
input.grad.zero_()
```
would do that.

Reply:
in that case we still need to store the gradient of input somewhere before we run a different graph.